source/openqasm/parser/antlr/qasm_parser.py

```diff
@@ -0,0 +1,926 @@
+    AssignmentOperator,
+    BinaryExpression,
+    BinaryOperator,
+    BitType,
+    BitTypeName,
+    BooleanLiteral,
+    Box,
+    BranchingStatement,
+    BreakStatement,
+    CalibrationDefinition,
+    CalibrationGrammarDeclaration,
+    ClassicalArgument,
+    ClassicalAssignment,
+    ClassicalDeclaration,
+    ComplexType,
+    Concatenation,
+    ConstantDeclaration,
+    Constant,
+    Concatenation,
+    ConstantDeclaration,
+    ConstantName,
+    ContinueStatement,
+    DelayInstruction,
+    DoubleDesignatorType,
+    DoubleDesignatorTypeName,
+    DurationOf,
+    EndStatement,
+    ExpressionStatement,
+    ExternDeclaration,
+    ForInLoop,
+    FunctionCall,
+    GateModifierName,
+    Identifier,
+    IndexExpression,
+    IntegerLiteral,
+    IODeclaration,
+    IOIdentifierName,
+    NoDesignatorType,
+    NoDesignatorTypeName,
+    OpenNode,
+    Program,
+    QuantumArgument,
+    QuantumGate,
+    QuantumGateModifier,
+    QuantumMeasurementAssignment,
+    QuantumPhase,
+    Qubit,
+    QubitDeclaration,
+    QubitDeclTypeName,
+    QuantumGateDefinition,
+    QuantumBarrier,
+    QuantumMeasurement,
+    QuantumReset,
+    RangeDefinition,
+    RealLiteral,
+    ReturnStatement,
+    Selection,
+    SingleDesignatorType,
+    SingleDesignatorTypeName,
+    Slice,
+    Span,
+    SubroutineDefinition,
+    Subscript,
+    StringLiteral,
+    TimeUnit,
+    DurationLiteral,
+    UnaryExpression,
+    UnaryOperator,
+    WhileLoop,
+)
+
+
+def parse(openqasm3_program: str) -> OpenNode:
+    lexer = qasm3Lexer(InputStream(openqasm3_program))
+    stream = CommonTokenStream(lexer)
+    parser = qasm3Parser(stream)
+
+    tree = parser.program()
+
+    return OpenNodeVisitor().visitProgram(tree)
+
+
+def get_span(node: Union[ParserRuleContext, TerminalNode]) -> Span:
+    """Get the span of a node"""
+    if isinstance(node, ParserRuleContext):
+        return Span(node.start.line, node.start.column, node.stop.line, node.stop.column)
+    else:
+        return Span(node.symbol.line, node.symbol.start, node.symbol.line, node.symbol.stop)
+
+
+def add_span(open_node: OpenNode, span: Span) -> OpenNode:
+    """Set the span of a node and return the node"""
+    open_node.span = span
+    return open_node
+
+
+def combine_span(first: Span, second: Span):
+    """Combine two spans and return the combined one"""
+    return Span(first.start_line, first.start_column, second.start_line, second.start_column)
+
+
+def span(func):
+    """Function decorator to automatic attach span to nodes for visit* methods."""
+
+    def wrapped(*args, **kwargs):
+        span = get_span(args[1])  # args[1] is ctx
+        node = func(*args, **kwargs)
+        if node is None:
+            raise ValueError(f"None encountered at {span}")
+        else:
+            node.span = span
+            return node
+
+    return wrapped
+
+
+class OpenNodeVisitor(qasm3Visitor):
+    @span
+    def visitProgram(self, ctx: qasm3Parser.ProgramContext):
```

Comment:
The current implementation does not handle `include*`. The first version will not support it, right?

Reply:
We imagine there will be a pre-processing step before this and all `include` will be expanded.

Reply:
Parsing is necessary to expand `include` and I think this AST is useful to do it.

Reply:
> Parsing is necessary to expand `include`

Agree. For parsing `include` though, I feel it's more efficient to just search for include and replace it with the file without constructing an AST.

Reply:
I just added `Include` as an AST node. Since the parser can parse a program in a string, we don't want the parser to resolve `include` from file system. The visitor of the AST can resolve the file and run another parser to parse the include file.